How do we handle compiler warnings with gcc9 for the 9.1 release?

As indicated in #8015 we did have a series of warnings with gcc9. After the fix #8034 I started to look into the issues reported via bundled software. Most of the warnings are of this kind (see also [1] for 300 of those warnings), now the one I found in our code:

```
/.../dealii_git/source/lac/petsc_parallel_sparse_matrix.cc: In member function ‘PetscScalar dealii::PETScWrappers::MPI::SparseMatrix::matrix_norm_square(const dealii::PETScWrappers::MPI::Vector&) const’:
/.../dealii_git/source/lac/petsc_parallel_sparse_matrix.cc:679:19: warning: implicitly-declared ‘dealii::PETScWrappers::MPI::Vector::Vector(const dealii::PETScWrappers::MPI::Vector&)’ is deprecated [-Wdeprecated-copy]
  679 |       Vector tmp(v);

/.../dealii_git/include/deal.II/lac/petsc_vector.h:277:7: note: because ‘dealii::PETScWrappers::MPI::Vector’ has user-provided ‘dealii::PETScWrappers::MPI::Vector& dealii::PETScWrappers::MPI::Vector::operator=(const dealii::PETScWrappers::MPI::Vector&)’
  277 |       operator=(const Vector &v);
      |       ^~~~~~~~
```

The warnings go both ways, namely complaints about classes that have a copy constructor but no copy assignment, and classes that have a copy assignment but not copy constructor. Many of them are located in the bundled boost-1.62.0. There is one in muparser and some in TBB (`enumerable_thread_specific`).

The question is how we handle this kind of warning for the release. My suggestion would be to disable the respective gcc warning `-Wno-deprecated-copy`. We could potentially fix them (I have boost and muparser things addressed locally, but not yet TBB).

The second kind of warning, which I have not yet addressed, is 
```
/home/kronbichler/sw/dealii_git/bundled/boost-1.62.0/include/boost/geometry/index/detail/varray_detail.hpp:544:13: warning: ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type ‘struct std::pair<dealii::Point<3, double>, unsigned int>’ with no trivial copy-assignment; use copy-assignment or copy-initialization instead [-Wclass-memaccess]
  544 |     ::memcpy(boost::addressof(*pos), boost::addressof(v), sizeof(V));
```
This is because we end up in a path for `memcpy` when we have a `std::pair` of `dealii::Point,uint` that is not trivially copyable. Again, we could probably address this by modifications in boost, but it is not trivial and I have not yet found where boost decides that the type in question would be trivially copyable.

As we have discussed during the last release in #6618, we might need to live with some warnings for new compilers. Yet, the situation is a bit frustrating due to the huge numbers of warnings that hides any useful error message, so I guess some action would help.

[1] https://cdash.kyomu.43-1.org/viewBuildError.php?type=1&buildid=9756

--- a/include/deal.II/base/config.h.in
+++ b/include/deal.II/base/config.h.in
@@ -374,6 +374,7 @@ _Pragma("GCC diagnostic ignored \"-Wunknown-warning-option\"")    /*!*/ \
 _Pragma("GCC diagnostic ignored \"-Wunknown-warning\"")           /*!*/ \
 _Pragma("GCC diagnostic ignored \"-Wextra\"")                     /*!*/ \
 _Pragma("GCC diagnostic ignored \"-Waddress-of-packed-member\"")        \
+_Pragma("GCC diagnostic ignored \"-Wdeprecated-copy\"")                 \
 _Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")         \
 _Pragma("GCC diagnostic ignored \"-Wexpansion-to-defined\"")            \
 _Pragma("GCC diagnostic ignored \"-Wexpansion-to-defined\"")            \

--- a/include/deal.II/base/graph_coloring.h
+++ b/include/deal.II/base/graph_coloring.h
@@ -24,8 +24,10 @@
 
 #  include <deal.II/lac/sparsity_tools.h>
 
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
 #  include <boost/unordered_map.hpp>
 #  include <boost/unordered_set.hpp>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
 
 #  include <functional>
 #  include <set>

--- a/include/deal.II/base/patterns.h
+++ b/include/deal.II/base/patterns.h
@@ -30,11 +30,13 @@
 
 #include <deal.II/fe/component_mask.h>
 
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
 #include <boost/archive/basic_archive.hpp>
 #include <boost/core/demangle.hpp>
 #include <boost/property_tree/ptree_fwd.hpp>
 #include <boost/property_tree/ptree_serialization.hpp>
 #include <boost/serialization/split_member.hpp>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
 
 #include <array>
 #include <deque>

--- a/include/deal.II/base/point.h
+++ b/include/deal.II/base/point.h
@@ -22,7 +22,9 @@
 #include <deal.II/base/exceptions.h>
 #include <deal.II/base/tensor.h>
 
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
 #include <boost/geometry.hpp>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
 
 #include <cmath>
 

--- a/include/deal.II/base/thread_local_storage.h
+++ b/include/deal.II/base/thread_local_storage.h
@@ -19,9 +19,11 @@
 
 #  include <deal.II/base/config.h>
 
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
 #  ifdef DEAL_II_WITH_THREADS
 #    include <tbb/enumerable_thread_specific.h>
 #  endif
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
 
 
 

--- a/include/deal.II/base/thread_management.h
+++ b/include/deal.II/base/thread_management.h
@@ -32,6 +32,7 @@
 #  include <tuple>
 #  include <utility>
 #  include <vector>
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
 #  ifdef DEAL_II_WITH_THREADS
 #    include <thread>
 #    ifdef DEAL_II_USE_MT_POSIX
@@ -40,6 +41,7 @@
 #    include <tbb/task.h>
 #    include <tbb/tbb_stddef.h>
 #  endif
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
 
 
 

--- a/include/deal.II/base/utilities.h
+++ b/include/deal.II/base/utilities.h
@@ -40,6 +40,8 @@
 #  endif
 #endif
 
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
+
 #include <boost/archive/binary_iarchive.hpp>
 #include <boost/archive/binary_oarchive.hpp>
 #include <boost/core/demangle.hpp>
@@ -53,6 +55,8 @@
 #  include <boost/iostreams/filtering_stream.hpp>
 #  include <boost/iostreams/stream.hpp>
 #endif
+
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
 
 DEAL_II_NAMESPACE_OPEN
 

--- a/include/deal.II/matrix_free/matrix_free.templates.h
+++ b/include/deal.II/matrix_free/matrix_free.templates.h
@@ -40,7 +40,9 @@
 #ifdef DEAL_II_WITH_THREADS
 #  include <deal.II/base/parallel.h>
 
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
 #  include <tbb/concurrent_unordered_map.h>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
 #endif
 
 #include <fstream>

